base/reduce.jl

```diff
@@ -61,7 +63,75 @@ function _foldl_impl(op::OP, init, itr) where {OP}
     return v
 end
 
-struct _InitialValue end
+@inline function _foldl_impl(op::OP, init, array::AbstractArray) where {OP}
+    if IndexStyle(array) isa IndexLinear
+        return invoke(_foldl_impl, Tuple{Any,Any,Any}, op, init, array)
```

Comment:
That we have to use `invoke` here is a codesmell for me. Why can't this be done with a dispatch on the IndexStyle?

Reply:
I just simply renamed the default implementation a56c37196484bdf148b3e44128636b2586e83a6f. I can wrap it with a new function with `IndexStyle` argument but I think this is simpler. What do you think?

The reason why I don't think `IndexStyle` is a good approach here is that this trait is insufficient for supporting more complex collections such as sparse arrays.